File: src/main/java/com/kishlaly/ta/model/SymbolData.java
package com.kishlaly.ta.model;

import com.kishlaly.ta.model.indicators.Indicator;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class SymbolData {

    public String symbol;
    public Timeframe timeframe;
    public List<QuoteJava> quotes;
    public Map<Indicator, List<? extends AbstractModelJava>> indicators = new HashMap<>();

    public QuoteJava getLastQuote() {
        return quotes.get(quotes.size() - 1);
    }

    public QuoteJava getPreLastQuote() {
        return quotes.get(quotes.size() - 2);
    }

    public SymbolData copy() {
        SymbolData copy = new SymbolData();
        copy.symbol = symbol;
        copy.timeframe = timeframe;
        copy.quotes = new ArrayList<>(quotes);
        copy.indicators = new HashMap<>(indicators);
        return copy;
    }

    public void clear() {
        quotes.clear();
        indicators.clear();
    }

}

File: src/main/kotlin/com/kishlaly/ta/model/SymbolData.kt
package com.kishlaly.ta.model

import com.kishlaly.ta.model.indicators.Indicator

data class SymbolData(
    val symbol: String,
    val timeframe: Timeframe,
    var quotes: List<Quote>,
    var indicators: MutableMap<Indicator, List<AbstractModel>> = mutableMapOf()
) {

    val lastQuote: Quote
        get() {
            return quotes.last()
        }

    val preLastQuote: Quote
        get() {
            return quotes[quotes.size - 2]
        }

    val quotesCount: Int
        get() {
            return quotes.size
        }

    val isEmptyQuotes: Boolean
        get() {
            return quotes.isEmpty()
        }

    val isEmptyIndicators: Boolean
        get() {
            return indicators.isEmpty()
        }

    fun clear() {
        indicators.clear()
    }

    fun indicator(indicator: Indicator) = indicators[indicator]

    infix fun quote(index: Int) = quotes[index]

    infix fun lastQuote(index: Int) = quotes[quotesCount - index]
}